### ISSUE TITLE
Dask Broadcast Outer Product Error

### ISSUE THREAD

**Issue by TNonet** [dask/dask#6009]
I hope I am not making a fool of myself, but I can't get the outer product [example](https://docs.dask.org/en/latest/array-api.html#dask.array.blockwise) using blockwise to work.

I have attempted to implement the following Example: 

Outer product multiplying x by y, two 1-d vectors
```python
z = blockwise(operator.mul, 'ij', x, 'i', y, 'j', dtype='f8')  # doctest: +SKIP
```

My first attempt was to use 1-D vectors. However, the result was not the outer product
```python
x = da.arange(10)
print(x.shape)
y = da.arange(10)
print(y.shape)
blockwise(da.multiply, 'ij', x, 'i', y, 'j', dtype='f8').compute()
```
```python
(10,)
(10,)
array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81])
```

My second attempt was to use 2-D vectors. However, this resulted in an Exception
The following do not work for me
```python
x = da.arange(10).reshape(-1, 1)
print(x.shape)
y = da.arange(10).reshape(1, -1)
print(y.shape)
blockwise(da.multiply, 'ij', x, 'i', y, 'j', dtype='f8').compute()
```
```python
(10, 1)
(1, 10)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-184-9afe940feddf> in <module>
      3 y = da.arange(10).reshape(1, -1)
      4 print(y.shape)
----> 5 blockwise(da.multiply, 'ij', x, 'i', y, 'j', dtype='f8').compute()

//anaconda3/lib/python3.7/site-packages/dask/array/blockwise.py in blockwise(func, out_ind, name, token, dtype, adjust_chunks, new_axes, align_arrays, concatenate, meta, *args, **kwargs)
    143 
    144     if align_arrays:
--> 145         chunkss, arrays = unify_chunks(*args)
    146     else:
    147         arginds = [(a, i) for (a, i) in toolz.partition(2, args) if i is not None]

//anaconda3/lib/python3.7/site-packages/dask/array/core.py in unify_chunks(*args, **kwargs)
   3167             )
   3168             if chunks != a.chunks and all(a.chunks):
-> 3169                 arrays.append(a.rechunk(chunks))
   3170             else:
   3171                 arrays.append(a)

//anaconda3/lib/python3.7/site-packages/dask/array/core.py in rechunk(self, chunks, threshold, block_size_limit)
   2169         from . import rechunk  # avoid circular import
   2170 
-> 2171         return rechunk(self, chunks, threshold, block_size_limit)
   2172 
   2173     @property

//anaconda3/lib/python3.7/site-packages/dask/array/rechunk.py in rechunk(x, chunks, threshold, block_size_limit)
    228         chunks = tuple(lc if lc is not None else rc for lc, rc in zip(chunks, x.chunks))
    229     chunks = normalize_chunks(
--> 230         chunks, x.shape, limit=block_size_limit, dtype=x.dtype, previous_chunks=x.chunks
    231     )
    232 

//anaconda3/lib/python3.7/site-packages/dask/array/core.py in normalize_chunks(chunks, shape, limit, dtype, previous_chunks)
   2420         raise ValueError(
   2421             "Chunks and shape must be of the same length/dimension. "
-> 2422             "Got chunks=%s, shape=%s" % (chunks, shape)
   2423         )
   2424     if -1 in chunks or None in chunks:

ValueError: Chunks and shape must be of the same length/dimension. Got chunks=((10,),), shape=(10, 1)
```
Different Combinations for the shape and chunks of x and y result in a similar error.

Dask Version:
'2.9.0'

I am unsure if this is an error with the example and thus I am implementing it wrong or if it is an error  with the blockwise operation itself.

**Comment by gforsyth** [comment]
Hey @TNonet -- thanks for reporting!  I think this is an error in the example. 
If you substitute `np.outer` in your `blockwise` example you get the correct results:

```
In [28]: x = da.arange(10)
    ...: print(x.shape)
    ...: y = da.arange(10)
    ...: print(y.shape)
(10,)
(10,)

In [29]: z = blockwise(np.outer, "ij", x, "i", y, "j").compute()

In [30]: z
Out[30]:
array([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0],
       [ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9],
       [ 0,  2,  4,  6,  8, 10, 12, 14, 16, 18],
       [ 0,  3,  6,  9, 12, 15, 18, 21, 24, 27],
       [ 0,  4,  8, 12, 16, 20, 24, 28, 32, 36],
       [ 0,  5, 10, 15, 20, 25, 30, 35, 40, 45],
       [ 0,  6, 12, 18, 24, 30, 36, 42, 48, 54],
       [ 0,  7, 14, 21, 28, 35, 42, 49, 56, 63],
       [ 0,  8, 16, 24, 32, 40, 48, 56, 64, 72],
       [ 0,  9, 18, 27, 36, 45, 54, 63, 72, 81]])
```

**Comment by TNonet** [comment]
Hi @gforsyth -- I appreciate the response. 

In this situation, when would one prefer 
```python
blockwise(np.outer, "ij", x, "i", y, "j")
```
over 
```python
np.outer(x, y)
```
assuming  x and y are dask vectors.

Or is it overkill to use blockwise here and the example is to just get people used to using it for more complex custom operations later?

**Comment by gforsyth** [comment]
Hey @TNonet -- since the introduction of the `__array_function__` NEP, I don't think there's a difference between those. 
`dask.array.outer`, in fact, calls `blockwise` that way (after inferring the output `dtype` and `flatten`ing both input arrays).

So yes, I think it's here just as an example of how `blockwise` works, and as a guide for those who want to implement their own array operations.  Would you like to put in a PR to fix the example?

**Comment by TNonet** [comment]
Hey @gforsyth -- I will add it to my todo list. I will also check the other examples. It shouldn't take much time.

**Comment by gforsyth** [comment]
Hey @TNonet --checking in to see if you're still interested in fixing this example in the docs?

**Comment by TNonet** [comment]
Hey @gforsyth . I appreciate the reminder. It is definitely still on my radar and I hope to get to it this week. With all of the current chaos, I had to shuffle some priorities.

**Comment by gforsyth** [comment]
> With all of the current chaos, I had to shuffle some priorities.

of course, no worries!

**Comment by TNonet** [comment]
I have begun looking over all the examples in blockwise. However, I believe I have potentially found another issue in [New Dimension Example](https://github.com/dask/dask/blob/5d5b43b4373677ff5fb7cf25eb4bdfb514cf3e76/dask/array/blockwise.py#L95).

```python
def f(x):
    return x[:, None] * np.ones((1, 5))
z1 = blockwise(f, 'az', x, 'a', new_axes={'z': True}, dtype=x.dtype)
z2 = blockwise(f, 'az', x, 'a', new_axes={'z': 5}, dtype=x.dtype)
z3 = blockwise(f, 'az', x, 'a', new_axes={'z': -1}, dtype=x.dtype)

for z in [z1, z2, z3]:
    print('z.shape pre compute():', z.shape)
    print('z.shape post compute():',z.compute().shape)
```

```python
z.shape pre compute(): (10, 1)
z.shape post compute(): (10, 5)
z.shape pre compute(): (10, 5)
z.shape post compute(): (10, 5)
z.shape pre compute(): (10, -1)
z.shape post compute(): (10, 5)
```

To me, this seems to show that the new_axes kwarg is not being used correctly. Or is this a situation where we assume the user is knowledgeable and won't try a pathological example?

I would normally expect an exception to be thrown here, but this shape issue is potentially not discoverable until it is too late.

If this is the case, I would suggest that this example might be difficult to generalize from as the '5' is hardcoded in both locations even though 'new_axes={'z': 5}' is not influential.

**Comment by TNonet** [comment]
Hey @gforsyth, I am wondering if I could have someone with a bit more experience comment on the efficacy of the above examples.

**Comment by gforsyth** [comment]
Hey @TNonet -- I'm far from an expert here, but a few thoughts:

The negative newaxis argument seems like it should be an error to me, that may merit a separate issue.  That said, I do think `blockwise` is expecting users to act in a non-pathological way.

As for the shape mismatching, if you first look at the `chunks` attribute of your `z1` and `z2` arrays, I think it's doing the right thing (creating a new axis that is 1 chunk wide of some size `z`). 

Whether it has to rechunk to complete the rest of the operations depends on whether that chunk width makes sense (I think). I would imagine that on larger dask arrays, the dimension of `new_axes` would have an impact on performance, but I'm speculating at this point.

**Comment by gforsyth** [comment]
> The negative newaxis argument seems like it should be an error to me, that may merit a separate issue.

sorry, disregard that.  `-1` chunksize indicates no chunking along that dimension

**Comment by martindurant** [comment]
Ping here - can we come to a resolution here?

**Comment by TNonet** [comment]
@martindurant Thanks for the ping!

I think the simplest solution would be to update the example as @gforsyth suggested. 

The other potential issue I raised seems to be something we can leave alone/for another day.

**Comment by martindurant** [comment]
Good to head. Please mark your PR as resolving this issue.

**Comment by TNonet** [comment]
It seems as if someone already fixed this issue #7119 
Thanks @bmerry
